File: DeepAtropos/trainOctantWithPriorsModel.R
library( ANTsR )
library( ANTsRNet )
library( keras )
library( tensorflow )
library( ggplot2 )

keras::backend()$clear_session()

Sys.setenv( "CUDA_VISIBLE_DEVICES" = "3" )

baseDirectory <- '/home/ntustison/Data/SixTissueSegmentation/'
scriptsDirectory <- paste0( baseDirectory, 'Scripts/' )
priorsDirectory <- paste0( baseDirectory, "Data/" )
source( paste0( scriptsDirectory, 'batchOctantWithPriorsGenerator.R' ) )

templateDirectory <- '/home/ntustison/Data/BrainAge2/Data/Templates/'
template <- antsImageRead( paste0( templateDirectory, "croppedMNI152.nii.gz" ) )
patchSize <- c( 112L, 112L, 112L )

priorFiles <- list.files( path = priorsDirectory, pattern = "croppedMniPriors",
			  recursive = FALSE, full.names = TRUE )

cat( "Readin gpriors\n" )
priors <- list()
for( i in seq.int( length( priorFiles ) ) )
  {
  priors[[i]] <- antsImageRead( priorFiles[i] )
  }


################################################
#
#  Create the model and load weights
#
################################################

cat( "Create u-net\n" )

classes <- c( 0:6 ) 
numberOfClassificationLabels <- length( classes )
imageModalities <- c( "T1", "background", "csf", "gm", "wm", "deepGm", "brainStem", "cerebellum" )
channelSize <- length( imageModalities )

unetModel <- createUnetModel3D( c( patchSize, channelSize ),
   numberOfOutputs = numberOfClassificationLabels, mode = 'classification',
   numberOfLayers = 4, numberOfFiltersAtBaseLayer = 16, dropoutRate = 0.0,
   convolutionKernelSize = c( 3, 3, 3 ), deconvolutionKernelSize = c( 2, 2, 2 ),
   weightDecay = 1e-5, additionalOptions = c( "attentionGating" ) )

cat( "HERE0\n" )


brainWeightsFileName <- paste0( scriptsDirectory, "/sixTissueOctantWithPriorsSegmentationWeights.h5" )
if( file.exists( brainWeightsFileName ) )
  {
  load_model_weights_hdf5( unetModel, filepath = brainWeightsFileName )
  # } else {
  # stop( "Weights file doesn't exist.\n" )  
  }

# weighted_loss <- weighted_categorical_crossentropy( weights = c( 0.05, 1, 2, 2, 3, 2, 2 ) )
# weighted_loss <- weighted_categorical_crossentropy( weights = c( 0.05, 1, 1, 2, 3, 2, 2 ) )
# weighted_loss <- weighted_categorical_crossentropy( weights = c( 0.05, 2, 1, 3, 4, 3, 3 ) )
# way too much csf --- weighted_loss <- weighted_categorical_crossentropy( weights = c( 0.05, 4, 1, 5, 5, 3, 3 ) )
# slightly too much csf weighted_loss <- weighted_categorical_crossentropy( weights = c( 0.05, 2.5, 1, 4, 4, 3, 3 ) )
# weighted_loss <- weighted_categorical_crossentropy( weights = c( 0.05, 2, 1, 5, 5, 3, 3 ) )
weighted_loss <- weighted_categorical_crossentropy( weights = c( 0.05, 1.5, 1, 3, 4, 3, 3 ) )

# weighted_loss <- weighted_categorical_crossentropy( weights = c( 0.0025, 0.045, 0.021, 0.026, 0.26, 0.56, 0.08 ) )

metric_multilabel_dice_coefficient <-
  custom_metric( "multilabel_dice_coefficient",
    multilabel_dice_coefficient )

loss_dice <- function( y_true, y_pred ) {
   -multilabel_dice_coefficient(y_true, y_pred)
   }
attr(loss_dice, "py_function_name") <- "multilabel_dice_coefficient"

unetModel %>% compile(
  optimizer = optimizer_adam(),
  loss = weighted_loss # tensorflow::tf$keras$losses$CategoricalCrossentropy(),
) #  metrics = c( metric_multilabel_dice_coefficient, metric_categorical_crossentropy ) )


################################################
#
#  Load the brain data
#
################################################

cat( "Loading brain data.\n" )

brainImages <-  
  Sys.glob( "/raid/data_NT/CorticalThicknessData2014/*/ThicknessAnts/*xMniInverseWarped.nii.gz" )


trainingImageFiles <- c()
trainingBrainMaskFiles <- c()

pb <- txtProgressBar( min = 0, max = length( brainImages ), style = 3 )
for( i in seq_len( length( brainImages ) ) )
  {
  setTxtProgressBar( pb, i )

  image <- brainImages[i]
  mask <- gsub( "InverseWarped", "BrainSegmentation", image )

  if( ! file.exists( image ) || ! file.exists( mask ) )
    {
    next  
    # stop( "Mask doesn't exist." )  
    }

  # # Do a quick check  
  # maskImage <- antsImageRead( mask )
  # if( sum( maskImage ) < 1000 )
  #   {
  #   cat( "mask: ", mask, "\n" )  
  #   next  
  #   }
  # imageArray <- as.array( antsImageRead( image ) )
  # if( any( is.na( imageArray ) ) || ( sum( imageArray ) <= 1000 ) )
  #   {
  #   cat( "image: ", image, "\n" )  
  #   next  
  #   }
    
  trainingImageFiles <- append( trainingImageFiles, image )
  trainingBrainMaskFiles <- append( trainingBrainMaskFiles, mask )
  }
cat( "\n" )  

cat( "Total training image files: ", length( trainingImageFiles ), "\n" )

cat( "\nTraining\n\n" )

###
#
# Set up the training generator
#

batchSize <- 12L 

# Split trainingData into "training" and "validation" componets for
# training the model.

numberOfData <- length( trainingImageFiles )
sampleIndices <- sample( numberOfData )

validationSplit <- floor( 0.8 * numberOfData )
trainingIndices <- sampleIndices[1:validationSplit]
numberOfTrainingData <- length( trainingIndices )
validationIndices <- sampleIndices[( validationSplit + 1 ):numberOfData]
numberOfValidationData <- length( validationIndices )

###
#
# Run training
#

track <- unetModel %>% fit_generator(
  generator = batchGenerator( batchSize = batchSize,
                              patchSize = patchSize, 
                              template = template,
			      priors = priors,
                              images = trainingImageFiles[trainingIndices],
                              segmentationImages = trainingBrainMaskFiles[trainingIndices],
			      segmentationLabels = classes,
                              doRandomContralateralFlips = FALSE,
                              doDataAugmentation = FALSE
                            ),
  steps_per_epoch = 100L,
  epochs = 200L,
  validation_data = batchGenerator( batchSize = batchSize,
                                    patchSize = patchSize, 
                                    template = template,
				    priors = priors,
                                    images = trainingImageFiles[validationIndices],
                                    segmentationImages = trainingBrainMaskFiles[validationIndices],
				    segmentationLabels = classes,
                                    doRandomContralateralFlips = FALSE,
                                    doDataAugmentation = FALSE
                                  ),  
  validation_steps = 40L,
  callbacks = list(
    callback_model_checkpoint( brainWeightsFileName,
      monitor = 'val_loss', save_best_only = TRUE, save_weights_only = TRUE,
      verbose = 1, mode = 'auto' ), 
     callback_reduce_lr_on_plateau( monitor = 'val_loss', factor = 0.3,
       verbose = 1, patience = 10, mode = 'auto' ),
    callback_early_stopping( monitor = 'val_loss', min_delta = 0.001,
      patience = 20 )
  )
)

save_model_weights_hdf5( unetModel, filepath = brainWeightsFileName )

File: DeepAtropos/Version0/train_octant_with_priors_model.py
import ants
import antspynet

import os
os.environ["CUDA_VISIBLE_DEVICES"] = "2"
import glob

import random
import math

import tensorflow as tf
import tensorflow.keras as keras
import tensorflow.keras.backend as K

from batch_octant_with_priors_generator import batch_generator

K.clear_session()
gpus = tf.config.experimental.list_physical_devices("GPU")
if len(gpus) > 0:
    tf.config.experimental.set_memory_growth(gpus[0], True)

tf.compat.v1.disable_eager_execution()

base_directory = '/home/ntustison/Data/SixTissueSegmentation/'
scripts_directory = base_directory + 'Scripts/'
priors_directory = base_directory + "Data/"

template_directory = '/home/ntustison/Data/BrainAge2/Data/Templates/'
template = ants.image_read(template_directory + "croppedMNI152.nii.gz")
patch_size = (112, 112, 112)

priors = list()
for i in range(7):
    priors.append(ants.image_read(priors_directory + "croppedMniPriors" + str(i) + ".nii.gz"))

################################################
#
#  Create the model and load weights
#
################################################

classes = list(range(7))
number_of_classification_labels = len(classes)
image_modalities = ["T1", "cerebellum"]
channel_size = len(image_modalities)

unet_model = antspynet.create_unet_model_3d((*patch_size, channel_size),
    number_of_outputs=number_of_classification_labels, mode="classification",
    number_of_layers=4, number_of_filters_at_base_layer=16, dropout_rate=0.0,
    convolution_kernel_size=(3, 3, 3), deconvolution_kernel_size=(2, 2, 2),
    weight_decay=1e-5, 
    additional_options=("attentionGating",))

brain_weights_filename = scripts_directory + "sixTissueOctantWithPriorsSegmentationWeights.h5"

#########
#
#          DO NOT LOAD PREVIOUS WEIGHTS!!!!!!!!!!!!!!
#
#
# if os.path.exists(brain_weights_filename):
#     unet_model.load_weights(brain_weights_filename)

unet_loss = antspynet.weighted_categorical_crossentropy(weights=(0.05, 1.5, 1, 3, 4, 3, 3 ))
unet_loss = antspynet.weighted_categorical_crossentropy(weights=(0.05, 1.5, 1, 3, 4, 3, 0.1 ))
unet_loss = antspynet.weighted_categorical_crossentropy(weights=(0.05, 3, 1, 3, 4, 3, 0.1 ))
unet_loss = antspynet.weighted_categorical_crossentropy(weights=(0.05, 10, 1, 3, 4, 3, 1 ))
unet_loss = antspynet.weighted_categorical_crossentropy(weights=(0.05, 10, 1, 3, 4, 3, 0.5 )) # pretty good.  Just a tad too much WM relative to GM and DeepGM
unet_loss = antspynet.weighted_categorical_crossentropy(weights=(0.05, 10, 1, 2, 4, 3, 0.5 ))

# unet_loss = antspynet.multilabel_surface_loss(dimensionality=3)

dice_loss = antspynet.multilabel_dice_coefficient(dimensionality=3, smoothing_factor=0.)
surface_loss = antspynet.multilabel_surface_loss(dimensionality=3)

def combined_loss(alpha):
    def combined_loss_fixed(y_true, y_pred):
        return (alpha * dice_loss(y_true, y_pred) +
                (1 - alpha) * surface_loss(y_true, y_pred))
    return(combined_loss_fixed)

wmh_loss = combined_loss(0.5)

unet_model.compile(optimizer=keras.optimizers.Adam(),
                   loss=unet_loss,
                   metrics=['accuracy', dice_loss])

##############################################
#
# Transfer weights
#
##############################################

unet_model_no_priors = \
    antspynet.create_unet_model_3d((*patch_size, 1),
    number_of_outputs=number_of_classification_labels, mode="classification",
    number_of_layers=4, number_of_filters_at_base_layer=16, dropout_rate=0.0,
    convolution_kernel_size=(3, 3, 3), deconvolution_kernel_size=(2, 2, 2),
    weight_decay=1e-5,
    additional_options=("attentionGating",))
unet_model_no_priors.load_weights(antspynet.get_pretrained_network("sixTissueOctantBrainSegmentation")) 

for i in range(2, len(unet_model.layers)):
    unet_model.get_layer(index=i).set_weights(unet_model_no_priors.get_layer(index=i).get_weights())

################################################
#
#  Load the data
#
################################################

print("Loading brain data.")

t1_images = glob.glob("/raid/data_NT/CorticalThicknessData2014/*/ThicknessAnts/*xMniInverseWarped.nii.gz")

training_t1_files = list()
training_seg_files = list()

for i in range(len(t1_images)):
    t1 = t1_images[i]
    seg = t1.replace("InverseWarped", "BrainSegmentation")

    training_t1_files.append(t1)
    training_seg_files.append(seg)

print("Total training image files: ", len(training_t1_files))

print( "Training")

###
#
# Set up the training generator
#

batch_size = 12

# Split trainingData into "training" and "validation" componets for
# training the model.

number_of_data = len(training_t1_files)
sample_indices = random.sample(range(number_of_data), number_of_data)

validation_split = math.floor(0.8 * number_of_data)

training_indices = sample_indices[:validation_split]
number_of_training_data = len(training_indices)

sampled_training_t1_files = list()
sampled_training_seg_files = list()

for i in range(number_of_training_data):
    sampled_training_t1_files.append(training_t1_files[training_indices[i]])
    sampled_training_seg_files.append(training_seg_files[training_indices[i]])

validation_indices = sample_indices[validation_split:]
number_of_validation_data = len(validation_indices)

sampled_validation_t1_files = list()
sampled_validation_seg_files = list()

for i in range(number_of_validation_data):
    sampled_validation_t1_files.append(training_t1_files[validation_indices[i]])
    sampled_validation_seg_files.append(training_seg_files[validation_indices[i]])


track = unet_model.fit_generator(
   generator=batch_generator(batch_size=batch_size,
                             patch_size=patch_size,
                             template=template,
                             priors=priors,
                             images=sampled_training_t1_files,
                             segmentation_images=sampled_training_seg_files,
                             segmentation_labels=classes,
                             do_random_contralateral_flips=True,
                             do_histogram_intensity_warping=False,
                             do_add_noise=False,
                             do_data_augmentation=False
                            ),
    steps_per_epoch=48,
    epochs=256,
    validation_data=batch_generator(batch_size=batch_size,
                                    patch_size=patch_size,
                                    template=template,
                                    priors=priors,
                                    images=sampled_training_t1_files,
                                    segmentation_images=sampled_training_seg_files,
                                    segmentation_labels=classes,
                                    do_random_contralateral_flips=True,
                                    do_histogram_intensity_warping=False,
                                    do_add_noise=False,
                                    do_data_augmentation=False
                                   ),
    validation_steps=24,
    callbacks=[
        keras.callbacks.ModelCheckpoint(brain_weights_filename, monitor='val_loss',
            save_best_only=True, save_weights_only=True, mode='auto', verbose=1),
        keras.callbacks.ReduceLROnPlateau(monitor='val_loss', factor=0.5,
           verbose=1, patience=10, mode='auto'),
        keras.callbacks.EarlyStopping(monitor='val_loss', min_delta=0.001,
           patience=20)
       ]
    )

unet_model.save_weights(brain_weights_filename)



